articles/testing-documentation-snippets/article.md

```diff
@@ -0,0 +1,173 @@
+
+> *From wikipedia :* Expect is an automation and non-regression testing tool written by Don Libes as an extension to the Tcl scripting language to test interactive applications like telnet, ftp, passwd, fsck, rlogin, ssh, or others.
+
+Running tests is simple, we give to expect a command to execute, also a string expected and it can do assertion on the result of the command.
+
+![schema expect](./images/schema-expect.png)
+
+
+For writing these tests, we opted for a YAML file in front of the snippets files. It simple to write, simple to read and simple to parse. They have to describe the test, give an expected string and the template we want to use. We also imagined a system of hooks (before and after) like in every test framework to execute some script for cleaning or other operations.
+
+```yaml
+name: kuzzle-connect
+description: Connects the SDK to Kuzzle
+hooks:
+  before: 
+  after: 
+template: default
+expected: Successfully connected
+```
+
+I found a cool Node.js package on NPM named [Nexpect](https://github.com/nodejitsu/nexpect). Then, I wrote a Node.js script for running tests for each language. For an easy implementation of every languages I make an abstract Tester class and I implement a children class for each language with lint and run command. In this way if we want to add another language for our SDKs it's very easy to just create a class and extend it to the mother tester class and only override commands to be executed for this new language.
+
+Here, the abstract class
+
+```js
+class Tester {
+
+  constructor () {
+    if (new.target === Tester) {
+      throw new TypeError("Cannot construct Tester instances directly");
+    }
+  }
+  
+  execLintCommand () {
+    // return a promise will executing 
+    // this.lintCommand added in childrens class
+  }
+  
+  execRunCommand () {
+    // return a promise will executing 
+    // this.runCommand added in childrens class
+  }
+  
+  // [...]
+}
+```
+
+And Here, the extended class for testing snippets in Go
+
+```js
+class GoTester extends Tester {
+  
+  constructor() {
+    super();
+    this.language = 'go';
+    this.runCommand = 'go run';
+    this.lintCommand = 'golint';
+  }
+
+}
+```
+
+And finally the last schema with the full process
+
+![schema4](./images/schema4.png)
+
+## Continuous integration and validation
+
+The last step and not the least was to put all of this process in a CI/CD workflow. At Kuzzle we use a Git-flow process. Each change in our documentation leads to a pull-request. And for each pull-request, we use Travis CI to run all containers for each language. I also add a report system to easily see which test was failed.
```

Comment:
"we use Travis CI to run all containers for each language" :point_down: "we run all the tests for each language in Travis CI"

"I also add a report system" :point_right: " I also added a report system"